trace-utils/src/msgpack_decoder/v04/decoder/mod.rs

```diff
@@ -0,0 +1,551 @@
+use rmp::{
+    decode,
+    decode::{read_array_len, RmpRead},
+    Marker,
+};
+use std::{collections::HashMap, f64};
+
+/// Decodes a slice of bytes into a vector of `TracerPayloadV04` objects.
+///
+///
+///
+/// # Arguments
+///
+/// * `data` - A mutable reference to a slice of bytes containing the encoded data.Bytes are
+///   expected to be encoded msgpack data containing a list of a list of v04 spans.
+///
+/// # Returns
+///
+/// * `Ok(Vec<TracerPayloadV04>)` - A vector of decoded `TracerPayloadV04` objects if successful.
+/// * `Err(DecodeError)` - An error if the decoding process fails.
+///
+/// # Errors
+///
+/// This function will return an error if:
+/// - The array length for trace count or span count cannot be read.
+/// - Any span cannot be decoded.
+///
+/// # Examples
+///
+/// ```
+/// use datadog_trace_protobuf::pb::Span;
+/// use datadog_trace_utils::msgpack_decoder::v04::decoder::from_slice;
+/// use rmp_serde::to_vec_named;
+///
+/// let span = Span {
+///     name: "test-span".to_owned(),
+///     ..Default::default()
+/// };
+/// let encoded_data = to_vec_named(&vec![vec![span]]).unwrap();
+/// let decoded_traces = from_slice(&mut encoded_data.as_slice()).expect("Decoding failed");
+///
+/// assert_eq!(1, decoded_traces.len());
+/// assert_eq!(1, decoded_traces[0].len());
+/// let decoded_span = &decoded_traces[0][0];
+/// assert_eq!("test-span", decoded_span.name);
+/// ```
+pub fn from_slice(data: &mut &[u8]) -> Result<Vec<TracerPayloadV04>, DecodeError> {
+    let trace_count = read_array_len(data).map_err(|_| {
+        DecodeError::InvalidFormat("Unable to read array len for trace count".to_owned())
+    })?;
+
+    let mut traces: Vec<TracerPayloadV04> = Default::default();
+
+    for _ in 0..trace_count {
+        let span_count = match read_array_len(data) {
+            Ok(count) => count,
+            Err(_) => {
+                return Err(DecodeError::InvalidFormat(
+                    "Unable to read array len for span count".to_owned(),
+                ))
+            }
+        };
+
+        let mut trace: Vec<Span> = Default::default();
+
+        for _ in 0..span_count {
+            let span = decode_span(data)?;
+            trace.push(span);
+        }
+        traces.push(trace);
+    }
+
+    Ok(traces)
+}
+
+#[inline]
+fn read_string_ref(buf: &[u8]) -> Result<(&str, &[u8]), DecodeError> {
+    decode::read_str_from_slice(buf).map_err(|e| match e {
+        DecodeStringError::InvalidMarkerRead(e) => DecodeError::InvalidFormat(e.to_string()),
+        DecodeStringError::InvalidDataRead(e) => DecodeError::InvalidConversion(e.to_string()),
+        DecodeStringError::TypeMismatch(marker) => {
+            DecodeError::InvalidType(format!("Type mismatch at marker {:?}", marker))
+        }
+        DecodeStringError::InvalidUtf8(_, e) => DecodeError::Utf8Error(e.to_string()),
+        _ => DecodeError::IOError,
+    })
+}
+
+#[inline]
+fn read_string(buf: &mut &[u8]) -> Result<String, DecodeError> {
```

Comment:
Can't this function just use `read_string_ref`, and then bump the `buf` like the code in `fill_span`?

Reply:
Yes, that works. Good idea!